### DISCUSSION TITLE
ssh-agent in wsl2 after restart wsl

### DISCUSSION THREAD

**Discussion by hmz22** [microsoft/WSL#11833] [Q&A]
Hello Dears;
I have problem with ssh-add in wsl2 I add this in .profile or .bashrc:
```
if [ -z "$SSH_AUTH_SOCK" ]; then
   # Check for a currently running instance of the agent
   RUNNING_AGENT="`ps -ax | grep 'ssh-agent -s' | grep -v grep | wc -l | tr -d '[:space:]'`"
   if [ "$RUNNING_AGENT" = "0" ]; then
        # Launch a new instance of the agent
        ssh-agent -s &> $HOME/.ssh/ssh-agent
   fi
   eval `cat $HOME/.ssh/ssh-agent` > /dev/null
   ssh-add $HOME/.ssh/<your ssh key> 2> /dev/null
fi
```
and it's worked for first time and after restart system or shutdown WSL and re run it ssh-add -l is empty and won't work and I must kill agent and re run WSL because the pid of agent changed and ssh-agent file data don't change e.g pid when run first time is 3412 and when restart WSL and run again pid is 213 but content of ssh-agent file is like this:
```
SSH_AUTH_SOCK=/tmp/ssh-XdsUS8eespqk/agent.3411; export SSH_AUTH_SOCK;
SSH_AGENT_PID=3412; export SSH_AGENT_PID;
echo Agent pid 3412;
```
can someone help me for fix it
I need it because I use vscode and devcontainer

**Comment by ademczuk** [comment]
## Resolution: Inline image rendering is working! (Feb 27, 2026)

The full implementation is now working end-to-end. PR #23764 has been updated with the complete fix.

### What was needed beyond the original TUI changes

The original analysis in this discussion correctly identified the 5 root causes. The TUI changes (tool-execution.ts, tui-event-handlers.ts, tui-formatters.ts, tui-session-actions.ts) handle detection and rendering. However, two additional upstream fixes were needed:

1. **Image tool MEDIA: emission** — The `image` tool must emit `MEDIA:/path/to/file` text blocks in its return content array. Without this, the tool result only contains description text and no file path for the TUI to render.

2. **Gateway MEDIA bypass** — The gateway's event handler strips `data.result` from tool events when verbose != "full" (the default). A targeted bypass preserves `data.result` when it contains MEDIA: image paths.

### Key architecture insight

The OpenClaw dist uses Rollup code splitting. The `reply-*.js` module contains its own copy of:
- `handleToolExecutionEnd` (the tool result event emitter)
- `emitAgentEvent` / `onAgentEvent` (the event bus)
- The image tool's `execute` function

The `pi-embedded-*.js` files contain *duplicate* copies that are used by different entry points. For the gateway's embedded agent runs, the **reply module's copies** are the ones that execute. This was confirmed via debug logging.

### Screenshot

![Working inline image](https://raw.githubusercontent.com/ademczuk/openclaw/feat/tui-inline-image-rendering/openclaw-tui-inline-cat.png)

See PR #23764 for the full implementation.
